### DISCUSSION TITLE
fetchSockets not working over redis adapter

### DISCUSSION THREAD

**Discussion by djllap** [socketio/socket.io#3977] [Q&A] [answered]
I've recently upgraded to 4.x, replacing socket.io-redis custom hook functions with `fetchSockets`. According to [this issue](https://github.com/socketio/socket.io-redis-adapter/issues/370#issuecomment-797812548) in the redis adapter github, `fetchSockets` should work with multiple nodes, but in my application, it is only returning sockets connected directly to the same socket.io instance. The redis adapter seems to be working when emitting events.

I'm using socket.io/redis-adapter 7.0.0 and socket.io 4.1.2

`socket.io server`
```
io = require('socket.io')(server, {transports: ['websocket']});
const {createAdapter} = require('@socket.io/redis-adapter');
const Redis = require('ioredis');

const pubClient = new Redis(config.redis);
const subClient = new Redis(config.redis);
io.adapter(createAdapter(pubClient, subClient));
```
`fetchSockets use in server code`
```
const sockets = await io.in(room).fetchSockets();
```

Is `fetchSockets` not able to pull socket instances from other nodes? Or is it maybe the `io.in(room)` that's limiting the results. If so, how would I go about getting sockets from a specific room?

**Comment by darrachequesne** [comment, marked as answer]
Yes, the Redis adapter supports the `fetchSockets`, and should return the socket instances from the other nodes.

I am not able to reproduce the issue: https://github.com/socketio/socket.io-fiddle/tree/multiple-servers

I tested with 2 servers and 2 clients (1 per server), and the following code seems to work:

```js
setInterval(async () => {
  const sockets = await io.fetchSockets();
  console.log("fetchSockets", sockets.length);
}, 5000);
```

Could you please check?

**Reply by djllap** [reply]
After some testing, it seems that `fetchSockets` is working correctly, but with and without specifying a room.

Still, when I run this code with 10 identical browser tabs, roughly half of them never receive the emit with a 2 node set up.
```
const sockets = await io.in(room).fetchSockets();
sockets.forEach(async (socket) => {
  if (socket.userId && socket.searchParams) {
      const user = await models.User.findByPk(socket.userId);
     {...}
      io.to(socket.id).emit(socketEvent, data);
  }
});
```

I'm just duplicating the tab, so the behavior should be the same. Maybe custom socket attributes (like `socket.userId`) don't carry over the adapter?

**Reply by darrachequesne** [reply]
> Maybe custom socket attributes (like socket.userId) don't carry over the adapter?

Yes indeed. You can attach custom attributes to the `data` attribute though:

```js
socket.data.userId = "...";
```

Reference: https://socket.io/docs/v4/server-instance/#fetchSockets

**Reply by djllap** [reply]
That was the problem alright. Thanks a lot for your help!
